specification/hugr.md

```diff
@@ -665,68 +672,103 @@ See [Type System](#type-system) for more on Resources.
 # may need some top level data, e.g. namespace?
 
 # Import other header files to use their custom types
+  # TODO: allow qualified, and maybe locally-scoped
 imports: [Quantum]
 
-# Declare custom types
-types:
-- name: QubitVector
-  # Opaque types can take type arguments, with specified names
-  args: [size]
-
-# Declare operations which aren't associated to a resource
-operations:
-- name: measure
-  description: "measure a qubit"
-  # We're going to implement this using ops defined in the "Quantum" resource
-  resource_reqs: [Quantum] 
-  inputs: [[null, Q]]
-  # the first element of each pair is an optional parameter name
-  outputs: [[null, Q], [measured, B]]
-
-# Declare some resource interfaces which provide the rest of the operations
 resources:
 - name: MyGates
+  # Declare custom types
+  types:
+  - name: QubitVector
+    # Opaque types can take type arguments, with specified names
+    args: [["size", u64]]
   operations:
+  - name: measure
+    description: "measure a qubit"
+    signature:
+      # No 'function' specified - use default type scheme interpreter.
+      # The first element of each pair is an optional parameter name.
+      inputs: [[null, Q]]  # Q is defined in Quantum resource
+      outputs: [[null, Q], ["measured", B]]
   - name: ZZPhase
     description: "Apply a parametric ZZPhase gate"
-    resource_reqs: [] # The "MyGates" resource will automatically be added as a requirement
-    inputs: [[null, Q], [null, Q], [angle, Angle]]
-    outputs: [[null, Q], [null, Q]]
+    signature:  # using default type scheme interpreter
+      inputs: [[null, Q], [null, Q], ["angle", Angle]]
+      outputs: [[null, Q], [null, Q]]
     misc:
       # extra data that may be used by some compiler passes
+      # and is passed to try_lower and compute_signature
       equivalent: [0, 1]
       basis: [Z, Z]
   - name: SU2
     description: "One qubit unitary matrix"
-    resource_reqs: []
-    inputs: [[null, Q]]
-    outputs: [[null, Q]]
+    args: # per-node values passed to the type-scheme interpreter, but not used in signature
+      - matrix: Opaque(complex_matrix,2,2)
+    signature:  # using default type scheme interpreter
+      inputs: [[null, Q]]
+      outputs: [[null, Q]]
+  - name: MatMul
+    description: "Multiply matrices of statically-known size"
+    args:  # per-node values passed to type-scheme-interpreter and used in signature
+      - i: U64
+      - j: U64
+      - k: U64
+    signature:  # using default type scheme interpreter
+      inputs: [["a", Array<i>(Array<j>(F64))], ["b", Array<j>(Array<k>(F64))]]
+      outputs: [[null, Array<i>(Array<k>(F64))]]
+  - name: max_float
+    description: "Variable number of inputs"
+    args:
+      - n: U64
+    signature:
+      function: with_repeats  # Name used to look up compute_signature impl in registry
```

Comment:
Perhaps the default type scheme interpreter allows this; or "with_repeats" is already registered but still needs to be used explicitly?